pkg/agent/run.go

```diff
@@ -184,37 +184,61 @@ func gatherData(ctx context.Context, config Config) []*api.DataReading {
 		dataGatherers[dgConfig.Name] = dg
 	}
 
-	// Fetch from all datagatherers
-	now := time.Now()
+	//TODO Change backoff parameters to those desired
+	backOff := backoff.NewExponentialBackOff()
+	backOff.MaxElapsedTime = 30 * time.Second
 	readings := []*api.DataReading{}
-	failedDataGatherers := []string{}
-	for k, dg := range dataGatherers {
-		i, err := dg.Fetch()
-		if err != nil {
-			log.Printf("Error fetching with DataGatherer %q: %s", k, err)
-			failedDataGatherers = append(failedDataGatherers, k)
-			continue
-		}
-
-		log.Printf("Gathered data for %q:\n", k)
+	completedDataGatherers := make(map[string]bool)
 
-		readings = append(readings, &api.DataReading{
-			ClusterID:    config.ClusterID,
-			DataGatherer: k,
-			Timestamp:    api.Time{Time: now},
-			Data:         i,
-		})
+	// Fetch from all datagatherers
+	getReadings := func() error {
+		var fullError error
+		for k, dg := range dataGatherers {
+			if completedDataGatherers[k] {
+				continue
+			}
+			dgData, err := dg.Fetch()
+			if err != nil {
+				if fullError == nil {
+					fullError = fmt.Errorf("%s: %v", k, err)
+				} else {
+					fullError = fmt.Errorf("%v\n%s: %v", fullError, k, err)
+				}
```

Comment:
Instead of building the string in this complicated way, I suggest that you replace `fullError error` with `dgErrors map[string]error`, where the key is the different `k`s.

So in case of error you just need to add a new element to the map.

Then, in the end, you can create a new error and format the message as you prefer.

Reply:
I see we are already using `go-multierror` in the project. In that case, it's best to just use it: https://github.com/hashicorp/go-multierror.

Reply:
Will do :)